cmd/k8s-operator/logger.go

```diff
@@ -0,0 +1,41 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+//go:build !plan9
+
+package main
+
+import (
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	kzap "sigs.k8s.io/controller-runtime/pkg/log/zap"
+	"tailscale.com/types/logger"
+)
+
+type (
+	logfSink struct {
+		logf logger.Logf
+	}
+)
+
+// wrapZapCore returns a zapcore.Core implementation that splits the core chain using zapcore.NewTee. This causes
+// logs to be simultaneously written to both the original core and the provided logger.Logf function.
+func wrapZapCore(core zapcore.Core, logf logger.Logf) zapcore.Core {
+	// We use a tee logger here so that logs are written to stdout/stderr normally while at the same time being
+	// sent upstream.
+	return zapcore.NewTee(core, zapcore.NewCore(&kzap.KubeAwareEncoder{
+		Encoder: zapcore.NewJSONEncoder(zap.NewProductionEncoderConfig()),
+		Verbose: true,
+	}, &logfSink{logf: logf}, zap.LevelEnablerFunc(func(lvl zapcore.Level) bool {
+		return true
+	})))
+}
+
+func (l *logfSink) Write(p []byte) (n int, err error) {
```

Comment:
The `logtail.Logger` type actually already implements the `io.Writer` interface that this implements. So compared to using that directly, we end up with an extra buffer allocation here inside `logf`'s `fmt.Fprintf`. To reduce the layers, and for increased logging efficiency, is it worth considering a function like this on tsnet's Server instead of `Logtailf`?:

```go
func (s *Server) LogtailWriter() io.Writer {
	return s.logtail
}
```

Reply:
That's a good point. It made more sense when this was originally implementing its own core.